src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala

```diff
@@ -532,41 +584,103 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
+            val Jacobian = DenseMatrix.zeros[Double](3, 3)
+            // Translate vertex coordinates from global to local where center of mass is local origin
+            val vertex1 = DenseVector[Double](vertices(face)(0)(0) - centerX,
+              vertices(face)(0)(1) - centerY,
+              vertices(face)(0)(2) - centerZ)
+            val vertex2 = DenseVector[Double](vertices(face)(vertexIndex + 1)(0) - centerX,
+              vertices(face)(vertexIndex + 1)(1) - centerY,
+              vertices(face)(vertexIndex + 1)(2) - centerZ)
+            val vertex3 = DenseVector[Double](vertices(face)(vertexIndex + 2)(0) - centerX,
+              vertices(face)(vertexIndex + 2)(1) - centerY,
+              vertices(face)(vertexIndex + 2)(2) - centerZ)
+
+            // Initialize Jacobian
+            Jacobian(0, 0) = vertex1(0)
+            Jacobian(0, 1) = vertex1(1)
+            Jacobian(0, 2) = vertex1(2)
+            Jacobian(1, 0) = vertex2(0)
+            Jacobian(1, 1) = vertex2(1)
+            Jacobian(1, 2) = vertex2(2)
+            Jacobian(2, 0) = vertex3(0)
+            Jacobian(2, 1) = vertex3(1)
+            Jacobian(2, 2) = vertex3(2)
+
+            val JacobianDet = linalg.det(Jacobian)
+
+            // Calculate x, y and z centroid increments
+            ( (vertex1(0) + vertex2(0) + vertex3(0)) * JacobianDet,
+              (vertex1(1) + vertex2(1) + vertex3(1)) * JacobianDet,
+              (vertex1(2) + vertex2(2) + vertex3(2)) * JacobianDet
+              )
+          } else {
+            (0.0, 0.0, 0.0)
+          }
+        }
+      }
+
+      val (centroidX, centroidY, centroidZ) = increments.fold (0.0, 0.0, 0.0) {
+        case ((centX1, centY1, centZ1),  (centX2, centY2, centZ2)) =>
+          (centX1 + centX2, centY1 + centY2, centZ1 + centZ2)
+      }
+
+      Array(centroidX / (volume * 24.0) + centerX,
+        centroidY / (volume * 24.0) + centerY,
+        centroidZ / (volume * 24.0) + centerZ)
     }
   }
 
   /**
-    * Calculates the volume of the block
-    * See http://wwwf.imperial.ac.uk/~rn/centroid.pdf for theoretical background.
+    * Calculates the volume of the block using simplex integration
     *
     * @return The volume of the block
     */
   private def findVolume: Double = {
-    val vertices = findVertices
-    val mesh = meshFaces(vertices)
+    val vertices = orientedVertices
 
     val volIncrements = faces.flatMap { face =>
-      mesh(face).map { m =>
-        // Access triangulation entries backwards so they are in counterclockwise order
-        val a_vals = vertices(face)(m(2))
-        val b_vals = vertices(face)(m(1))
-        val c_vals = vertices(face)(m(0))
-        val a = DenseVector[Double](a_vals(0), a_vals(1), a_vals(2))
-        val b = DenseVector[Double](b_vals(0), b_vals(1), b_vals(2))
-        val c = DenseVector[Double](c_vals(0), c_vals(1), c_vals(2))
-
-        val n = linalg.cross(b - a, c - a)
-        a dot n
+      vertices(face).map { faceVertex =>
```

Comment:
The comments from the `centroid` method may also apply here: initializing the Jacobian more cleanly and using `dropRight` to avoid the complex logic based on indexes.